sds/src/scanner/mod.rs

```diff
@@ -457,6 +471,9 @@ impl<'a, E: Encoding> ContentVisitor<'a> for ScannerContentVisitor<'a, E> {
         let mut group_data: AHashMap<TypeId, Box<dyn Any>> = AHashMap::new();
 
         rule_visitor.visit_rule_indices(|rule_index| {
+            if self.blocked_rules.contains(&rule_index) {
```

Comment:
For each rule and each leaf of the event, this method will be called. It could be expensive for big events.
I think it can be optimised a bit if this is pushed to the `RuleIndexVisitor`, `ActiveRuleTree` to drop the rules earlier in the scan pipeline. I'm not sure it's possible but I wonder if we can perform the exclusion of rule once per event and not one per leaf. Did you explore this option?

Reply:
Yes this is a tricky part. 
This is not super optimal as we will hash every rule_index visited for each content. 
Alternative maybe what you suggested but not sure how to do that, or faster structure search (like bool array of excluded rules)

Reply:
I will dig into it a bit more

Reply:
I create a Vec version so that inference is O(1) and there is no hashing involved. Not sure how to do better using the ScopedRuleset. Maybe @fuchsnj you have a better idea ?

Reply:
A `Vec` is fine as long as the list stays small (<10 or so), after that we probably need something else. The `ScopedRuleset` wouldn't support this very well, since the "excluded" rules in there are still scanned for Multipass V0.

Does the `blocked_rules` really change frequently enough to need to pass it into _every_ scan function? Can we just re-build the scanner with the new rules whenever this list changes?

Reply:
The perf of the `Vec` is _probably_ fine. It seems like this will usually be an empty `Vec`, so the `.contains()` will be fast. It's only populated during a slow rollout, so that should be somewhat temporary.

Reply:
An alternative to re-building the entire scanner, we could add a function to the scanner to disable rules, which could re-build just the `ScopedRuleset` (instead of having to re-compile all of the regexes).

Reply:
We can't rebuild the scanner since the slow rollout has a rand part in it, meaning it can change between 2 scans and so we would have to rebuild the scanner very often. 
This is not a contains anymore but a Vec<bool> indexed by rule_idx containing true when the rule is blocked so access is always fast whatever the number of blocked rules. Downside is that it takes more memory for potentially nothing.

Reply:
Okay, the `Vec` is probably fine. I would just make sure that the common case of not having a rollout at all doesn't have any significant overhead (don't allocate an array of `false`) for each scan.

Just brainstorming here, but an alternative may be to have multiple `ScopedRuleset`'s, and you can dynamically choose which one to use for each scan. The `Vec` is probably fine though and we can look at perf improvements if they are needed.

Reply:
Ok I move to a pure vector + contains search.
When the vector is populated it will be small and transient.